### ISSUE TITLE
Speedup by direct allocation in focus loss

### ISSUE THREAD

**Issue by rmcavoy** [signatrix/efficientdet#30]
The below lines with torch.ones and torch.zeros in the focus loss are allocating massive arrays on the home device and then transferring them to the gpu using .cuda. This is inefficient especially in cases where DataParallel is over larger numbers of gpu's as the main process will have to move the data to each gpu individually.

```

            # compute the loss for classification
            targets = torch.ones(classification.shape) * -1

            if torch.cuda.is_available():
                targets = targets.cuda()

             alpha_factor = torch.ones(targets.shape) * alpha

            if torch.cuda.is_available():
                alpha_factor = alpha_factor.cuda()

            zeros = torch.zeros(cls_loss.shape)
            if torch.cuda.is_available():
                  zeros = zeros.cuda()
```

Instead the data should be allocated directly on the gpu using e.g. ones_like (below) to copy all traits of the original array including their device. My tests with the below on my private fork show an approximately 4x training speedup from this change alone when tested on 6 GPU's using DataParallel (though I am also using a custom parallel dataloader so your performance may vary)
```
targets = torch.ones_like(classification) * -1
alpha_factor = torch.ones_like(targets) * alpha
zeros = torch.zeros_like(cls_loss))
```

**Comment by rmcavoy** [comment]
This change should also speedup training even on 1 gpu as the code will have fewer giant arrays to move around. Also I assume this problem is a feature of whatever everyone seems to be deriving their focus_loss from as I have seen this problem near identically in several different codes. It also makes the logic easier as you don't need to check if the cuda is available since it will just put the array wherever the original arrays are located.

**Comment by glenn-jocher** [comment]
@rmcavoy good recommendation. You might want to submit a PR directly.
